File: app/src/main/java/com/zeapo/pwdstore/utils/PasswordItem.java
package com.zeapo.pwdstore.utils;

import androidx.annotation.NonNull;
import com.zeapo.pwdstore.crypto.PgpActivity;

import java.io.File;

public class PasswordItem implements Comparable {

    public final static char TYPE_CATEGORY = 'c';
    public final static char TYPE_PASSWORD = 'p';

    private final char type;
    private final String name;
    private final PasswordItem parent;
    private final File file;
    private final String fullPathToParent;
    private final String longName;

    /**
     * Create a password item
     * <p>
     * Make it protected so that we use a builder
     */
    private PasswordItem(String name, PasswordItem parent, char type, File file, File rootDir) {
        this.name = name;
        this.parent = parent;
        this.type = type;
        this.file = file;
        fullPathToParent = file.getAbsolutePath()
                .replace(rootDir.getAbsolutePath(), "")
                .replace(file.getName(), "");
        longName = PgpActivity.getLongName(fullPathToParent, rootDir.getAbsolutePath(), toString());
    }

    /**
     * Create a new Category item
     */
    public static PasswordItem newCategory(String name, File file, PasswordItem parent, File rootDir) {
        return new PasswordItem(name, parent, TYPE_CATEGORY, file, rootDir);
    }

    /**
     * Create a new parentless category item
     */
    public static PasswordItem newCategory(String name, File file, File rootDir) {
        return new PasswordItem(name, null, TYPE_CATEGORY, file, rootDir);
    }

    /**
     * Create a new password item
     */
    public static PasswordItem newPassword(String name, File file, PasswordItem parent, File rootDir) {
        return new PasswordItem(name, parent, TYPE_PASSWORD, file, rootDir);
    }

    /**
     * Create a new parentless password item
     */
    public static PasswordItem newPassword(String name, File file, File rootDir) {
        return new PasswordItem(name, null, TYPE_PASSWORD, file, rootDir);
    }

    public char getType() {
        return this.type;
    }

    String getName() {
        return this.name;
    }

    public PasswordItem getParent() {
        return this.parent;
    }

    public File getFile() {
        return this.file;
    }

    public String getFullPathToParent() {
        return this.fullPathToParent;
    }

    public String getLongName() {
        return longName;
    }

    @Override
    public String toString() {
        return this.getName().replace(".gpg", "");
    }

    @Override
    public boolean equals(Object o) {
        // Makes it possible to have a category and a password with the same name
        return o != null
                && o.getClass() == PasswordItem.class
                && ((PasswordItem) o).getFile().equals(this.getFile());
    }

    @Override
    public int compareTo(@NonNull Object o) {
        PasswordItem other = (PasswordItem) o;
        // Appending the type will make the sort type dependent
        return (this.getType() + this.getName())
                .compareToIgnoreCase(other.getType() + other.getName());
    }
}

File: app/src/main/java/dev/msfjarvis/aps/data/password/PasswordItem.kt
/*
 * Copyright © 2014-2021 The Android Password Store Authors. All Rights Reserved.
 * SPDX-License-Identifier: GPL-3.0-only
 */
package dev.msfjarvis.aps.data.password

import android.content.Context
import android.content.Intent
import dev.msfjarvis.aps.data.repo.PasswordRepository
import dev.msfjarvis.aps.ui.crypto.BasePgpActivity
import dev.msfjarvis.aps.ui.main.LaunchActivity
import java.io.File

data class PasswordItem(
  val name: String,
  val parent: PasswordItem? = null,
  val type: Char,
  val file: File,
  val rootDir: File
) : Comparable<PasswordItem> {

  val fullPathToParent = file.absolutePath.replace(rootDir.absolutePath, "").replace(file.name, "")

  val longName = BasePgpActivity.getLongName(fullPathToParent, rootDir.absolutePath, toString())

  override fun equals(other: Any?): Boolean {
    return (other is PasswordItem) && (other.file == file)
  }

  override fun compareTo(other: PasswordItem): Int {
    return (type + name).compareTo(other.type + other.name, ignoreCase = true)
  }

  override fun toString(): String {
    return name.replace("\\.gpg$".toRegex(), "")
  }

  override fun hashCode(): Int {
    return 0
  }

  /** Creates an [Intent] to launch this [PasswordItem] through the authentication process. */
  fun createAuthEnabledIntent(context: Context): Intent {
    val intent = Intent(context, LaunchActivity::class.java)
    intent.putExtra("NAME", toString())
    intent.putExtra("FILE_PATH", file.absolutePath)
    intent.putExtra("REPO_PATH", PasswordRepository.getRepositoryDirectory().absolutePath)
    intent.action = LaunchActivity.ACTION_DECRYPT_PASS
    return intent
  }

  companion object {

    const val TYPE_CATEGORY = 'c'
    const val TYPE_PASSWORD = 'p'

    @JvmStatic
    fun newCategory(name: String, file: File, parent: PasswordItem, rootDir: File): PasswordItem {
      return PasswordItem(name, parent, TYPE_CATEGORY, file, rootDir)
    }

    @JvmStatic
    fun newCategory(name: String, file: File, rootDir: File): PasswordItem {
      return PasswordItem(name, null, TYPE_CATEGORY, file, rootDir)
    }

    @JvmStatic
    fun newPassword(name: String, file: File, parent: PasswordItem, rootDir: File): PasswordItem {
      return PasswordItem(name, parent, TYPE_PASSWORD, file, rootDir)
    }

    @JvmStatic
    fun newPassword(name: String, file: File, rootDir: File): PasswordItem {
      return PasswordItem(name, null, TYPE_PASSWORD, file, rootDir)
    }
  }
}
